File: 1452.people-whose-list-of-favorite-companies-is-not-a-subset-of-another-list.cs
/*
 * @lc app=leetcode id=1452 lang=csharp
 *
 * [1452] People Whose List of Favorite Companies Is Not a Subset of Another List
 */

// @lc code=start
public class Solution 
{
    public IList<int> PeopleIndexes(IList<IList<string>> favoriteCompanies) 
    {
        int n = favoriteCompanies.Count();
        List<HashSet<string>> preference = new List<HashSet<string>>();
        List<int> result = new List<int>();
        HashSet<int> skip = new HashSet<int>();
        int[] indices = new int[n];
        int[] preferenceSize = new int[n];

        for (int i = 0; i < n; i++)
        {
            preference.Add(new HashSet<string>(favoriteCompanies[i].ToList()));
            indices[i] = i;
            preferenceSize[i] = favoriteCompanies[i].Count;
        }

        Array.Sort(preferenceSize, indices);
        
        for (int i = 0; i < n-1; i++)
        {
            if (!skip.Contains(i))
            {
                int j = i + 1;
                while (j < n)
                {
                    if (preference[indices[i]].IsSubsetOf(preference[indices[j]]))
                    {
                        if (preference[indices[i]].Count == preference[indices[j]].Count)
                        {
                            skip.Add(j);
                        }
                        break;
                    }
                    else
                    {
                        j++;
                    }
                }
                if (j == n)
                {
                    result.Add(indices[i]);
                }
            }
        }

        if (!skip.Contains(n-1))
        {
            result.Add(indices[n-1]);
        }
        result.Sort();
        return result;
    }
}
// @lc code=end

File: 1452.people-whose-list-of-favorite-companies-is-not-a-subset-of-another-list.py
#
# @lc app=leetcode id=1452 lang=python3
#
# [1452] People Whose List of Favorite Companies Is Not a Subset of Another List
#

# @lc code=start
class Solution:
    # Use hash set to store each person's preference
    # Use issubset to check if each pair completely overlap

    def peopleIndexes(self, favoriteCompanies: List[List[str]]) -> List[int]:
        n = len(favoriteCompanies)
        preference = [set(choice) for choice in favoriteCompanies]
        sortedIndices = [x[0] for x in sorted(enumerate(preference), key = lambda x: len(x[1]))]
        result = set()
        skip = set()
        for i in range(n-1):
            if i not in skip:
                j = i+1
                while j < n:
                    if preference[sortedIndices[i]].issubset(preference[sortedIndices[j]]):
                        if len(preference[sortedIndices[i]]) == len(preference[sortedIndices[j]]):
                            skip.add(j)
                        break
                    else:
                        j += 1
                if j == n:
                    result.add(sortedIndices[i])
        if n-1 not in skip:
            result.add(sortedIndices[n-1])
        return sorted(list(result))
# @lc code=end

